// synced-json-tree/Tree.js
"use strict";

const LocalVersion = require(`./LocalVersion.js`);
const Version = require(`./Version.js`);

const Tree = class {

    constructor () {

        this.childTrees = new Map();

        this.localVersion = LocalVersion.oldest;

        this.pendingFuncs = [];

        this.version = Version.oldest;

    }

    build (Path) {

        let tree;

        for (tree of this.iterativelyBuild(Path)) {

        }

        return tree;

    }

    *iterativelyBuild (Path) {

        let tree = this;

        yield tree;

        for (const child of Path) {

            let childTree = tree.childTrees.get(child);

            if (childTree === undefined) {

                childTree = new Tree();

                tree.childTrees.set(child, childTree);

            }

            tree = childTree;

            yield tree;

        }

    }

    *Traversal () {

        yield this;

        for (const tree of this.childTrees.values()) {

            yield* tree.Traversal();

        }

    }

    Versions (Path) {

        const versions = new Array(1 + Path.length);

        let i = 0;

        let tree = this;

        while (true) {

            versions[i] = tree.version;

            if (i === Path.length) {

                break;

            }

            tree = tree.childTrees.get(Path[i]);

            i++;   

            if (tree === undefined) {

                versions.fill(Version.oldest, i);

                break;

            }

        }

        return versions;

    }

    };

module.exports = Tree;
